### PR TITLE
improved the error message when encounter overflow. This is just a te…

### PR DESCRIPTION
**Pull request by zhexuany**
…mp commit. Rebase is needed

This PR aims to improve error handling in `tidb`. Once PR merged into master, `tidb` will have same behavior as `mysql` does.

### CODE REVIEW

**Comment by zhexuany** [comment]
This pr address this issue. https://github.com/pingcap/tidb/issues/2862

**Comment by zhexuany** [comment]
The error message are generated in overflow.go which reside in util/type. These error message requires runtime information i.e. `select 10000000000 + 10000000000`. These two uint have to be shown in error message. There is no way that we can predefine a error message and use it at runtime. Hence, I propose creating error message at runtime and format the error message according to the runtime data.

**Comment by coocood** [comment]
The error code is much more important than the error message.
Most of the applications only check error code.
Formatted error message is hard to compare.

**Comment by coocood** [comment]
@zhexuany 
You can take a look at the package `terror` and how it is used.

**Comment by zhexuany** [comment]
@coocood gotcha. I think I already know a better way to address this issue.

File: util/types/overflow.go [hunk 1]
```diff
@@ -18,6 +18,15 @@ import (
 
 	"fmt"
 	"github.com/juju/errors"
+	"github.com/pingcap/tidb/terror"
+)
+
+const (
+	CodeUintAdditionOperationOverflow terror.ErrCode = 1
```
Comment (shenli) [inline util/types/overflow.go]: It is better to move this definition into util/types/errors.go. And if it is a mysql standard error, please also add the error code mapping in init(). Protocol layer will use this mapping to convert it to corresponding mysql error code.

**Comment by zhexuany** [comment]
Will push all commits to this PR tonight for reviewing.

File: util/types/overflow.go (hunk 1 above)
Comment (zhexuany) [inline util/types/overflow.go]: @shenli  this pr is not ready yet. I am still doing more work. It is my bad not saying this in this PR. As you can see, the todo list is still marked as todo.

**Comment by shenli** [comment]
@zhexuany OK

**Comment by zhexuany** [comment]
@coocood  
~~~
select 10000000000000000000 + 10000000000000000000;
~~~
generates the following error message. 
~~~
ERROR 1264 (22003): BIGINT UNSIGNED value is out of range in '(10000000000000000000, 10000000000000000000)'
~~~

Is this error message ok?

**Comment by coocood** [comment]
@zhexuany 
MySQL returns error code `1690`

**Comment by zhexuany** [comment]
@coocood 
~~~
 return 0, errors.Trace(mysql.NewErr(mysql.ErrDataOutOfRange, "BIGINT UNSIGNED", fmt.Sprintf("(%d, %d)", a, b)))
~~~
This snippet gives 
~~~
ERROR 1105 (HY000): ERROR 1690 (22003): BIGINT UNSIGNED value is out of range in '(10000000000000000000, 10000000000000000000)'
~~~

The error code is correct but in a wired structure. Is this correct?

**Comment by coocood** [comment]
@zhexuany 
`typesMySQLErrCodes` need to be updated.

**Comment by zhexuany** [comment]
@coocood I already did that. Actually, my question is the error info's structure. 

~~~
ERROR 1105 (HY000): ERROR 1690 (22003): 
~~~
The first one indicates the error happens in `types`.  the second one indicates the error happens in mysql overflow. 

Is this understanding correct?

**Comment by zhexuany** [comment]
@coocood @shenli PTAL

**Comment by coocood** [comment]
LGTM
Thank you for your contribution!

File: expression/builtin_math.go [hunk 2]
```diff
@@ -24,6 +24,7 @@ import (
 	"strconv"
 	"strings"
 
+	"fmt"
```
Comment (zimulala) [inline expression/builtin_math.go]: Please move this line to line-26.

File: util/types/overflow.go [hunk 3]
```diff
@@ -16,16 +16,14 @@ package types
 import (
 	"math"
 
+	"fmt"
```
Comment (zimulala) [inline util/types/overflow.go]: Move this line to line-18.

File: expression/builtin_math.go [hunk 4]
```diff
@@ -372,7 +373,12 @@ func (b *builtinPowSig) eval(row []types.Datum) (d types.Datum, err error) {
 	if err != nil {
 		return d, errors.Trace(err)
 	}
-	d.SetFloat64(math.Pow(x, y))
+
+	power := math.Pow(x, y)
+	if math.IsInf(power, -1) || math.IsInf(power, 1) || math.IsNaN(power) {
```
Comment (zimulala) [inline expression/builtin_math.go]: Please add some tests for this code.

Comment (zhexuany) [inline expression/builtin_math.go]: DONE.

File: util/types/overflow.go (hunk 3 above)
Comment (zhexuany) [inline util/types/overflow.go]: DONE

File: expression/builtin_math.go (hunk 2 above)
Comment (zhexuany) [inline expression/builtin_math.go]: DONE

**Comment by zimulala** [comment]
LGTM

**Pull request merged**